formats/mdm/MDM_2_4.ddl

```diff
@@ -0,0 +1,515 @@
+  First
+    -- Parse the message type octet followed by the number of messages
+    -- octet, which has a fixed value corresponding to each message
+    -- type.
+    Type_VRT                  = @Match [0x0, 0x1, 0x0, 0x1]
+    Type_Time_Of_Day          = @Match [0x0, 0x3, 0x0, 0x1]
+    Type_Signal_Port_User_ID  = @Match [0x0, 0x4, 0x0, 0x1]
+    Type_Health_Status        = @Match [0x0, 0x5, 0x0, 0x1]
+    Type_Command              = @Match [0x0, 0x6, 0x0, 0x1]
+    Type_Switch_Group_User_ID = @Match [0x0, 0x7, 0x0, 0x1]
+
+    Type_Acknowledgement = block
+      @Match [0x0, 0x2, 0x0, 0x0]
+      -- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification
+      --
+      -- In the Acknowledge messge type, the "ack" octet parsed in the
+      -- message header is treated as a response octet which takes one
+      -- of a number of meaningful values.
+      --
+      -- The specification states that all other possible values are
+      -- Reserved for Future Use; if they are encountered here, they
+      -- result in a parse failure.
+      response = raw_ack_byte as? MDM_Ack_Response_Value
+
+-- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification, Table
+-- 74
+bitdata MDM_Ack_Response_Value where
+  -- Acknowledged (A)
+  Resp_A = 0x01
+  -- Invalid Message (IM)
+  Resp_IM = 0x02
+  -- Invalid Function (IF)
+  Resp_IF = 0x04
+  -- Invalid Range (IR)
+  Resp_IR = 0x08
+  -- Resource Unavailable (RU)
+  Resp_RU = 0x10
+  -- Unexpected Sequence Number (US)
+  Resp_US = 0x20
+  -- Unauthorized User (UU)
+  Resp_UU = 0x40
+
+-- See 6.1.2.2, MDM Body
+def MDM_Body (ty: MDM_Type) =
+  case ty of
+    Type_VRT ->
+      -- See 6.3.1.2.1: VRT Message Format
+      -- VRT_Message
+      Fail "VRT message parsing not implemented"
+
+    -- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification
+    -- Acknowledgement messages have no body.
+    Type_Acknowledgement info ->
+      ^ {| Body_Acknowledgement = { } |}
+
+    -- See 6.3.3: MDM Time of Day Message (Type 3) Specification
+    Type_Time_Of_Day ->
+      {| Body_Time_Of_Day = BEUInt32 |}
+
+    -- See 6.3.4: MDM Signal Port User ID Message (Type 4) Specification
+    Type_Signal_Port_User_ID ->
+      Parse_Body_Signal_Port_User_ID
+
+    -- See 6.3.5: MDM Health Message (Type 5) Specification
+    Type_Health_Status ->
+      block
+        let f1 = BEUInt32 as? Health_Status_Field_1
+        let f2 = BEUInt32 as? Health_Status_Field_2
+        ^ {| Body_Health_Status = { status_field_1 = f1, status_field_2 = f2 } |}
+
+    -- See 6.3.6: MDM Command Message (Type 6) Specification
+    Type_Command ->
+      block
+        let command_field = BEUInt32 as? Command_Command_Field
+        let configuration_field = BEUInt32
+        let waveform_op_field = BEUInt32
+        ^ {| Body_Command = { command = command_field,
+                              configuration = configuration_field,
+                              waveform_operation = waveform_op_field } |}
```

Comment:
Perhaps rewrite this as:

```
def BodyCommand =
  block
    command = BEUInt32 as? Command_Command_Field
    configuration = BEUInt32
    waveform_operation = BEUInt32
```
then the last line can become just:
```
{| Body_Command = BodyCommand |}
```   
 I like this because it avoid the anonymous type, and avoids the `let`s

Reply:
Thanks, fixed in 226eb48c51788e5bed689fe2845e3b14ea66c0a8